src/graph.py

```diff
@@ -0,0 +1,122 @@
+"""A gragh data structure."""
+
+
+class Graph(object):
+    """Graph."""
+
+    def __init__(self):
+        """Create the dictionary that will be the graph."""
+        self.g = {}
+
+    def nodes(self):
+        """Return a list of all the nodes in the graph."""
+        nodes = []
+        for key in self.g:
+            nodes += [key]
+        return nodes
+
+    def edges(self):
+        """Return a list of all the edges in the graph."""
+        edges = []
+        for key in self.g:
+            for edge in self.g[key]:
+                edges += [(key, edge)]
+        return edges
+
+    def add_node(self, n):
+        """Add a node to the graph."""
+        self.g.setdefault(n, [])
+
+    def add_edge(self, n1, n2):
+        """Add an edge between two nodes."""
+        weight = 0
+        if type(n1) is int or type(n1) is float:
+            weight += n1
+        else:
+            weight += 1
+        if type(n2) is int or type(n2) is float:
+            weight += n2
+        else:
+            weight += 1
+        try:
+            self.g[n1] += [(n2, weight)]
+        except KeyError:
+                self.g[n1] = [(n2, weight)]
+
+    def del_node(self, n):
+        """Delete a node from the graph."""
+        try:
+            del self.g[n]
```

Comment:
So you delete the node from the graph, but you don't check to see that there aren't any nodes in the graph that are still pointing to the one that you delete.